pep-9999.rst

```diff
@@ -0,0 +1,501 @@
+PEP: 9999
+Title: tomllib: Support for parsing TOML in the Standard Library
+Author: Taneli Hukkinen, Shantanu Jain <hauntsaninja at gmail.com>
+Sponsor: Petr Viktorin <encukou@gmail.com>
+Discussions-To: https://discuss.python.org/t/adopting-recommending-a-toml-parser/4068
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 01-Jan-2022
+Python-Version: 3.11
+Post-History: 1900-01-01
+
+
+Abstract
+========
+
+This proposes adding a module, ``tomllib``, to the standard library for
+parsing TOML (Tom's Obvious Minimal Language,
+`https://toml.io <https://toml.io/en/>`_).
+
+
+Motivation
+==========
+
+The TOML format is the format of choice for Python packaging, as evidenced by
+:pep:`517`, :pep:`518` and :pep:`621`. Including TOML support in the standard
+library helps avoid bootstrapping problems for Python build tools. Currently
+most Python build tools need to vendor a TOML parsing library.
+
+Python tools are increasingly configurable via TOML, for examples: ``black``,
+``mypy``, ``pytest``, ``tox``, ``pylint``, ``isort``. Those that are not, such
+as ``flake8``, cite the lack of standard library support as a `main reason why
+<https://github.com/PyCQA/flake8/issues/234#issuecomment-812800657>`_.
+
+Given the special place TOML already has in the Python ecosystem, it makes sense
+for this to be an included battery.
+
+Finally, TOML as a format is increasingly popular (some reasons for this are
+outlined in PEP 518). Hence this is likely to be a generally useful addition,
+even looking beyond the needs of Python packaging and Python tooling: various
+Python TOML libraries have about 2000 reverse dependencies on PyPI. For
+comparison, ``requests`` has about 28k reverse dependencies.
+
+
+Rationale
+=========
+
+This PEP proposes basing the standard library support for reading TOML on the
+third party library ``tomli``
+(`github.com/hukkin/tomli <https://github.com/hukkin/tomli>`_).
+
+Many projects have recently switched to using ``tomli``, for example, ``pip``,
+``build``, ``pytest``, ``mypy``, ``black``, ``flit``, ``coverage``,
+``setuptools-scm``, ``cibuildwheel``.
+
+``tomli`` is actively maintained and well-tested. ``tomli`` is about 800 lines
+of code with 100% test coverage and passes all tests in a test suite `proposed as
+the official TOML compliance test suite <https://github.com/toml-lang/compliance/pull/8>`_,
+as well as `the more established BurntSushi/toml-test suite <https://github.com/BurntSushi/toml-test>`_.
+
+
+Specification
+=============
+
+A new module ``tomllib`` with the following functions will be added:
+
+.. code-block::
+
+   def load(fp: SupportsRead[bytes], /, *, parse_float: Callable[[str], Any] = float) -> dict[str, Any]: ...
+   def loads(s: str, /, *, parse_float: Callable[[str], Any] = float) -> dict[str, Any]: ...
```

Comment:
```suggestion
   def load(fp: SupportsRead[bytes], /, *, parse_float: Callable[[str], Any] = ...) -> dict[str, Any]: ...
   def loads(s: str, /, *, parse_float: Callable[[str], Any] = ...) -> dict[str, Any]: ...
```

I think the default argument is not relevant, should be considered an implementation detail, and seems to be a convention to elide it.

Reply:
I actually think it's more useful than the type annotation :)
If removed, the default behavior should be mentioned in the prose (even though it's straightforward).

Reply:
> If removed, the default behavior should be mentioned in the prose (even though it's straightforward).

Yeah I think I'd prefer doing this.

The reason why I'd want to consider this a non-annotated implementation detail is that I fear one day we may find out there is some special case where `float(<insert-special-TOML-float-here>)` does not return the appropriate result or even raises `ValueError`. If it happens it'd be best to set the default to `None` and make the `None` correspond to a slightly customized `float` callable.

Reply:
Ah, that makes sense!

Reply:
Hm, seeing the change to the text... The same argument can be made for the recommendation to use Decimal, can't it?

Reply:
Do you mean that `decimal.Decimal` could in theory fail to parse a TOML float? Yeah agreed. But at least that's only a recommendation, not baked into the API specification. Do you think we should change the wording? E.g. "a function that returns a `decimal.Decimal` can be used in cases where precision is important".

Reply:
Made this change and adjusted the wording of the `decimal.Decimal` bit!